text/0000-inline-buildpack.md

```diff
@@ -0,0 +1,126 @@
+
+We propose two new keys in the `[[build.buildpacks]]` table in `project.toml`:
+
+- `api` - the API compatibility version of the buildpack. Matches the schema of `buildpack.toml`
+- `inline` - the contents of a script that will be run as the `bin/build` of the inline buildpack
+- `shell` - defines the shebang line of the script in `inline`
+
+When any of these keys are provided the following keys are disallowed: `version`, `uri`.
+
+(Note: there is no need to implement `bin/detect`. The detect phase will always pass because there's no foreseeable reason that a user would create an inline buildpack that they don't want to use)
+
+# How it Works
+[how-it-works]: #how-it-works
+
+When an entry in the `[[build.buildpacks]]` table contains an `inline` value, the platform WILL NOT attempt to download the buildpack `id@version`. Instead, it will create an ephemeral buildpack, with the following:
+
+- a `bin/detect` that always passes (i.e. `exit 0`)
+- a `bin/build` that contains the value of `inline` script without any changes (i.e. no magic)
+- a `buildpack.toml` using the `id` and `version` from the buildpack entry in `project.toml`
+
+When the ephemeral buildpack is executed, its working directory will contain the artifacts in the app repo. In this way, an `inline` script MAY reference other scripts in the repo. For example, a `lib/utils.sh` may be sourced by an inline Bash script.
+
+## Example
+
+The following `project.toml` creates a buildpack that will run two Rake tasks after running JVM and Ruby buildpacks.
+
+```toml
+[project]
+id = "my-app"
+
+[[build.buildpacks]]
+id = "example/jvm"
+version = "1.0"
+
+[[build.buildpacks]]
+id = "example/ruby"
+version = "1.0"
+
+[[build.buildpacks]]
+api = "0.3"
+id = "me/rake-tasks"
+inline = """
+  rake war
+  rake db:migrate
+"""
+```
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+- It discourages buildpack reuse (i.e. if it's really easy to create a one-off buildpack, people are less like to make reuseable buildpacks and share them)
+- It encourages the copy-paste problems created by `Dockerfile`
+- There is no built-in mechanism to provide dependencies in the build plan. However, this can be worked-around with a generic build-plan-buildpack that either statically or dynamically creates entries in the build plan.
+
+# Alternatives
+[alternatives]: #alternatives
+
+- https://github.com/buildpacks/rfcs/pull/17
+- [Packfile](https://github.com/sclevine/packfile/)
+
+# Prior Art
+[prior-art]: #prior-art
+
+- `Dockerfile`
+- [Packfile](https://github.com/sclevine/packfile/)
+
+# Unresolved Questions
+[unresolved-questions]: #unresolved-questions
+
+- Can an inline buildpack be published to a registry?
+- Can inline buildpacks be root buildpacks?
+- Can a `project.toml` contain more than one inline buildpack?
+- How will platforms implement this behavior?
```

Comment:
Does this rfc assume that a platform will provide the inline buildpack as an ephemeral/temporary buildpack to the lifecycle?

Reply:
Or will the lifecycle need to become project.toml aware?

Reply:
Ultimately, `project.toml` is a platform concern. But I hope that in the future, lifecycle will provide some tools that make it easier for a platform to support `project.toml`. That means for now, a platform must handle this (and other project.toml features) on it's own and would need to provide the inline buildpack as an ephemeral/temporary buildpack to the lifecycle.

Reply:
If we specify an `inline` version of `buildpack.toml` we could solve both of these problems? The inline buildpack in `project.toml` could be tranforms into an "inline" `buildpack.toml`. https://github.com/buildpacks/rfcs/pull/86#discussion_r451643134